// @solid/access-control-policy/dist/mod.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.matchIssuer = exports.matchClient = exports.matchAgent = exports.match = exports.applyPolicy = exports.allowAccessModes = exports.MATCHER_ATTRIBUTES = exports.ACP = exports.ACL = exports.ACCESS_MODES = void 0;
// Constants
var access_modes_1 = require("./constant/access_modes");
Object.defineProperty(exports, "ACCESS_MODES", { enumerable: true, get: function () { return access_modes_1.ACCESS_MODES; } });
var acl_1 = require("./constant/acl");
Object.defineProperty(exports, "ACL", { enumerable: true, get: function () { return acl_1.ACL; } });
var acp_1 = require("./constant/acp");
Object.defineProperty(exports, "ACP", { enumerable: true, get: function () { return acp_1.ACP; } });
var matcher_attributes_1 = require("./constant/matcher_attributes");
Object.defineProperty(exports, "MATCHER_ATTRIBUTES", { enumerable: true, get: function () { return matcher_attributes_1.MATCHER_ATTRIBUTES; } });
// Algorithms
var allow_access_modes_1 = require("./algorithm/allow_access_modes");
Object.defineProperty(exports, "allowAccessModes", { enumerable: true, get: function () { return allow_access_modes_1.allowAccessModes; } });
var apply_policy_1 = require("./algorithm/apply_policy");
Object.defineProperty(exports, "applyPolicy", { enumerable: true, get: function () { return apply_policy_1.applyPolicy; } });
var match_1 = require("./algorithm/match");
Object.defineProperty(exports, "match", { enumerable: true, get: function () { return match_1.match; } });
var match_agent_1 = require("./algorithm/match_agent");
Object.defineProperty(exports, "matchAgent", { enumerable: true, get: function () { return match_agent_1.matchAgent; } });
var match_client_1 = require("./algorithm/match_client");
Object.defineProperty(exports, "matchClient", { enumerable: true, get: function () { return match_client_1.matchClient; } });
var match_issuer_1 = require("./algorithm/match_issuer");
Object.defineProperty(exports, "matchIssuer", { enumerable: true, get: function () { return match_issuer_1.matchIssuer; } });
