// opinionated-eslint-config/dist/stylistic.js
import stylisticPlugin from '@stylistic/eslint-plugin';
import antfuPlugin from 'eslint-plugin-antfu';
export default [{
        name: 'antfu/stylistic',
        plugins: {
            antfu: antfuPlugin,
            '@stylistic': stylisticPlugin,
        },
        rules: {
            ...stylisticPlugin.configs.customize({
                indent: 2,
                quotes: 'single',
                semi: true,
            }).rules,
            'antfu/consistent-list-newline': 'error',
            'antfu/curly': 'error',
            'antfu/if-newline': 'error',
            'antfu/top-level-function': 'error',
            '@stylistic/array-bracket-spacing': ['error', 'always', {
                    singleValue: true,
                    objectsInArrays: false,
                    arraysInArrays: false,
                }],
            // Conflicts with @stylistic/object-curly-spacing
            '@stylistic/block-spacing': 'off',
            '@stylistic/brace-style': ['error', '1tbs', { allowSingleLine: false }],
            '@stylistic/generator-star-spacing': ['error', { before: false, after: true }],
            '@stylistic/indent-binary-ops': 'error',
            '@stylistic/member-delimiter-style': ['error', {
                    multiline: { delimiter: 'semi', requireLast: true },
                    singleline: { delimiter: 'semi', requireLast: false },
                }],
            '@stylistic/no-extra-parens': ['error', 'all', {
                    // To prevent conflicts with @stylistic/no-mixed-operators
                    nestedBinaryExpressions: false,
                }],
            '@stylistic/object-curly-spacing': ['error', 'always', {
                    objectsInObjects: false,
                    arraysInObjects: false,
                }],
            '@stylistic/operator-linebreak': ['error', 'after'],
            '@stylistic/semi': ['error', 'always'],
            '@stylistic/semi-style': ['error', 'last'],
            '@stylistic/space-before-function-paren': ['error', {
                    anonymous: 'never',
                    named: 'never',
                    asyncArrow: 'never',
                    catch: 'always',
                }],
            '@stylistic/switch-colon-spacing': 'error',
            '@stylistic/quote-props': ['error', 'as-needed', {
                    keywords: false,
                    unnecessary: true,
                    numbers: false,
                }],
            '@stylistic/yield-star-spacing': ['error', 'after'],
        },
    }];
