// opinionated-eslint-config/dist/test.js
import jestPlugin from 'eslint-plugin-jest';
export default [{
        name: 'opinionated/test',
        files: ['test/**/*.ts'],
        plugins: {
            jest: jestPlugin,
        },
        languageOptions: {
            globals: jestPlugin.environments.globals.globals,
        },
        rules: {
            ...jestPlugin.configs.all.rules,
            // Rule is not smart enough to check called function in the test
            'jest/expect-expect': 'off',
            'jest/valid-title': ['error', {
                    mustNotMatch: {
                        describe: /\.$/u.source,
                    },
                    mustMatch: {
                        it: /\.$/u.source,
                    },
                }],
            // Default rules that are too drastic
            'jest/max-expects': 'off',
            'jest/no-conditional-in-test': 'off',
            'jest/no-hooks': 'off',
            'jest/padding-around-all': 'off',
            'jest/padding-around-expect-groups': 'off',
            'jest/prefer-ending-with-an-expect': 'off',
            'jest/prefer-expect-assertions': 'off',
            'jest/prefer-importing-jest-globals': 'off',
            'jest/prefer-lowercase-title': 'off',
            'jest/prefer-strict-equal': 'off',
            'jest/require-hook': 'off',
            '@typescript-eslint/naming-convention': 'off',
            '@typescript-eslint/no-confusing-void-expression': 'off',
            '@typescript-eslint/no-explicit-any': 'off',
            '@typescript-eslint/no-unsafe-argument': 'off',
            '@typescript-eslint/no-unsafe-assignment': 'off',
            '@typescript-eslint/no-unsafe-call': 'off',
            '@typescript-eslint/no-unsafe-member-access': 'off',
            '@typescript-eslint/no-unsafe-return': 'off',
            '@typescript-eslint/unbound-method': 'off',
            // Incorrectly detects usage of undefined in 'toHaveBeenLastCalledWith' checks
            'unicorn/no-useless-undefined': 'off',
        },
    }];
