// opinionated-eslint-config/dist/typescript.js
import typescriptPlugin from '@typescript-eslint/eslint-plugin';
import typescriptParser from '@typescript-eslint/parser';
import antfuPlugin from 'eslint-plugin-antfu';
export const GLOB_TS = ['**/*.?([cm])ts', '**/*.?([cm])tsx'];
export default [
    {
        name: 'opinionated/typescript/setup',
        files: GLOB_TS,
        plugins: {
            antfu: antfuPlugin,
            '@typescript-eslint': typescriptPlugin,
        },
        languageOptions: {
            parserOptions: {
                projectService: true,
                tsconfigRootDir: import.meta.dirname,
            },
            parser: typescriptParser,
        },
    },
    {
        name: 'opinionated/typescript/rules',
        files: GLOB_TS,
        rules: {
            ...typescriptPlugin.configs['eslint-recommended'].overrides[0].rules,
            ...typescriptPlugin.configs.strict.rules,
            ...typescriptPlugin.configs.stylistic.rules,
            'max-len': ['error', { code: 120, ignoreUrls: true }],
            'no-dupe-class-members': 'off',
            'no-loss-of-precision': 'off',
            'no-redeclare': 'off',
            'no-use-before-define': 'off',
            'no-useless-constructor': 'off',
            '@typescript-eslint/ban-ts-comment': ['error', { 'ts-expect-error': true }],
            '@typescript-eslint/consistent-type-definitions': 'off',
            '@typescript-eslint/consistent-type-imports': [
                'error',
                { disallowTypeAnnotations: false, prefer: 'type-imports' },
            ],
            '@typescript-eslint/explicit-member-accessibility': 'error',
            '@typescript-eslint/method-signature-style': 'error',
            '@typescript-eslint/no-dupe-class-members': 'error',
            '@typescript-eslint/no-dynamic-delete': 'off',
            '@typescript-eslint/no-empty-object-type': 'off',
            '@typescript-eslint/no-explicit-any': 'error',
            '@typescript-eslint/no-extraneous-class': ['error', {
                    allowConstructorOnly: false,
                    allowEmpty: false,
                    allowStaticOnly: false,
                }],
            '@typescript-eslint/no-import-type-side-effects': 'error',
            '@typescript-eslint/no-invalid-void-type': 'off',
            '@typescript-eslint/no-loss-of-precision': 'error',
            '@typescript-eslint/no-non-null-assertion': 'off',
            '@typescript-eslint/no-redeclare': 'error',
            '@typescript-eslint/no-require-imports': 'error',
            '@typescript-eslint/no-unsafe-function-type': 'error',
            '@typescript-eslint/no-unused-vars': 'off',
            '@typescript-eslint/no-use-before-define': ['error', { classes: false, functions: false, variables: true }],
            '@typescript-eslint/no-useless-constructor': 'off',
            '@typescript-eslint/no-wrapper-object-types': 'error',
            '@typescript-eslint/prefer-ts-expect-error': 'error',
            '@typescript-eslint/triple-slash-reference': 'off',
            '@typescript-eslint/unified-signatures': 'off',
            'unicorn/filename-case': ['error', {
                    cases: {
                        camelCase: true,
                        pascalCase: true,
                        kebabCase: false,
                        snakeCase: false,
                    },
                }],
        },
    },
    {
        name: 'opinionated/typescript/rules-type-aware',
        files: GLOB_TS,
        ignores: ['**/*.md'],
        rules: {
            ...typescriptPlugin.configs['strict-type-checked-only'].rules,
            ...typescriptPlugin.configs['stylistic-type-checked-only'].rules,
            'no-implied-eval': 'off',
            'no-throw-literal': 'off',
            '@typescript-eslint/consistent-type-assertions': ['error', { assertionStyle: 'as' }],
            '@typescript-eslint/explicit-function-return-type': ['error', {
                    allowExpressions: false,
                    allowTypedFunctionExpressions: false,
                    allowHigherOrderFunctions: false,
                }],
            '@typescript-eslint/naming-convention': [
                'error',
                {
                    selector: 'default',
                    format: ['camelCase'],
                    leadingUnderscore: 'forbid',
                    trailingUnderscore: 'forbid',
                },
                {
                    selector: 'import',
                    format: null,
                },
                {
                    selector: 'variable',
                    format: ['camelCase', 'UPPER_CASE'],
                    leadingUnderscore: 'forbid',
                    trailingUnderscore: 'forbid',
                },
                {
                    selector: 'typeLike',
                    format: ['PascalCase'],
                },
                {
                    selector: ['typeParameter'],
                    format: ['PascalCase'],
                    prefix: ['T'],
                },
            ],
            '@typescript-eslint/no-confusing-void-expression': ['error', { ignoreArrowShorthand: true }],
            '@typescript-eslint/no-floating-promises': ['error', { ignoreVoid: false }],
            // Sometimes required due to incorrect Record usage
            '@typescript-eslint/no-unnecessary-condition': 'off',
            // There are several cases where this is relevant, such as interfaces or making more specific return types
            '@typescript-eslint/no-unnecessary-type-parameters': 'off',
            '@typescript-eslint/no-unnecessary-qualifier': 'error',
            '@typescript-eslint/prefer-nullish-coalescing': ['error', { ignoreIfStatements: true }],
            '@typescript-eslint/prefer-readonly': 'error',
            '@typescript-eslint/promise-function-async': 'error',
            '@typescript-eslint/require-array-sort-compare': 'error',
            '@typescript-eslint/require-await': 'off',
            // Less strict defaults
            '@typescript-eslint/restrict-template-expressions': 'error',
            '@typescript-eslint/return-await': ['error', 'in-try-catch'],
            '@typescript-eslint/switch-exhaustiveness-check': 'error',
        },
    },
];
