// opinionated-eslint-config/dist/unicorn.js
import unicornPlugin from 'eslint-plugin-unicorn';
export default [{
        name: 'opinionated/unicorn',
        plugins: {
            unicorn: unicornPlugin,
        },
        rules: {
            'unicorn/better-regex': 'error',
            'unicorn/empty-brace-spaces': 'error',
            'unicorn/consistent-function-scoping': 'error',
            'unicorn/error-message': 'error',
            'unicorn/escape-case': 'error',
            'unicorn/expiring-todo-comments': ['error', {
                    ignoreDatesOnPullRequests: false,
                    terms: ['todo'],
                    allowWarningComments: false,
                }],
            'unicorn/explicit-length-check': 'error',
            'unicorn/new-for-builtins': 'error',
            'unicorn/no-array-for-each': 'error',
            'unicorn/no-array-reduce': 'error',
            'unicorn/no-await-in-promise-methods': 'error',
            'unicorn/no-for-loop': 'error',
            'unicorn/no-instanceof-array': 'error',
            'unicorn/no-invalid-remove-event-listener': 'error',
            'unicorn/no-lonely-if': 'error',
            'unicorn/no-negated-condition': 'error',
            'unicorn/no-nested-ternary': 'error',
            'unicorn/no-new-array': 'error',
            'unicorn/no-new-buffer': 'error',
            'unicorn/no-object-as-default-parameter': 'error',
            'unicorn/no-process-exit': 'error',
            'unicorn/no-thenable': 'error',
            'unicorn/no-useless-fallback-in-spread': 'error',
            'unicorn/no-useless-length-check': 'error',
            'unicorn/no-useless-promise-resolve-reject': 'error',
            'unicorn/no-useless-spread': 'error',
            'unicorn/no-useless-undefined': 'error',
            'unicorn/no-zero-fractions': 'error',
            'unicorn/number-literal-case': 'error',
            'unicorn/prefer-array-find': 'error',
            'unicorn/prefer-array-flat-map': 'error',
            'unicorn/prefer-array-index-of': 'error',
            'unicorn/prefer-array-some': 'error',
            'unicorn/prefer-at': 'error',
            'unicorn/prefer-code-point': 'error',
            'unicorn/prefer-date-now': 'error',
            'unicorn/prefer-default-parameters': 'error',
            'unicorn/prefer-dom-node-text-content': 'error',
            'unicorn/prefer-includes': 'error',
            'unicorn/prefer-math-trunc': 'error',
            'unicorn/prefer-native-coercion-functions': 'error',
            'unicorn/prefer-negative-index': 'error',
            'unicorn/prefer-node-protocol': 'error',
            'unicorn/prefer-number-properties': 'error',
            'unicorn/prefer-object-from-entries': 'error',
            'unicorn/prefer-optional-catch-binding': 'error',
            'unicorn/prefer-reflect-apply': 'error',
            'unicorn/prefer-regexp-test': 'error',
            'unicorn/prefer-set-has': 'error',
            'unicorn/prefer-set-size': 'error',
            'unicorn/prefer-spread': 'error',
            'unicorn/prefer-string-replace-all': 'error',
            'unicorn/prefer-string-slice': 'error',
            'unicorn/prefer-string-starts-ends-with': 'error',
            'unicorn/prefer-structured-clone': 'error',
            'unicorn/prefer-type-error': 'error',
            'unicorn/require-array-join-separator': 'error',
            'unicorn/require-number-to-fixed-digits-argument': 'error',
            'unicorn/throw-new-error': 'error',
            'unicorn/filename-case': ['error', {
                    cases: {
                        camelCase: false,
                        pascalCase: false,
                        kebabCase: true,
                        snakeCase: false,
                    },
                    ignore: [
                        // CODE_OF_CONDUCT.md, etc.
                        /[A-Z_]+\.md$/u,
                    ],
                }],
        },
    }];
